src/Pact/Native/Capabilities.hs

```diff
@@ -126,18 +129,19 @@ composeCapability =
     composeCapability' :: NativeFun e
     composeCapability' i [TApp app _] = gasUnreduced i [] $ do
       -- enforce in defcap
-      isDefCap <- uses evalCallStack $ preview (traverse . sfApp . _Just . _1 . faDefType . _Defcap)
-      void $ case isDefCap of
-        Nothing -> evalError' i "compose-capability valid only within defcap body"
-        Just {} -> do
-          -- should be granted in-module
-          granted <- evalCap True app
-          -- if newly granted, add to composed list
-          forM_ granted $ \newcap -> evalCapabilities . capComposed %= (newcap:)
+      defcapInStack >>= \p -> unless p $ evalError' i "compose-capability valid only within defcap body"
+      -- should be granted in-module
+      granted <- evalCap True app
+      -- if newly granted, add to composed list
+      forM_ granted $ \newcap -> evalCapabilities . capComposed %= (newcap:)
       return $ toTerm True
     composeCapability' i as = argsError' i as
 
 
+defcapInStack :: Eval e Bool
+defcapInStack = fmap isJust $ uses evalCallStack $ preview (traverse . sfApp . _Just . _1 . faDefType . _Defcap)
```

Comment:
So here we're traversing the call stack and looking to see if we're in a `defcap`?

Reply:
Yep, I'll add a comment